letsencrypt-auto-source/pieces/pipstrap.py

```diff
@@ -156,12 +156,13 @@ def main():
                                      temp,
                                      digest)
                      for path, digest in PACKAGES]
-        check_output('pip install --no-index --no-deps -U ' +
-                     # Disable cache since we're not using it and it otherwise
-                     # sometimes throws permission warnings:
-                     ('--no-cache-dir ' if has_pip_cache else '') +
-                     ' '.join(quote(d) for d in downloads),
-                     shell=True)
+        # On Windows, pip self-upgrade is not possible, it must be done through python interpreter.
+        command = ['pip'] if name != 'nt' else ['python', '-m', 'pip']
```

Comment:
I think we can move to `python -m pip` for all OSes. pip 1.2, which is the first that supported that invocation style, came out 7 years ago. Let's try that and make sure we run test_everything before we release, just to be sure we aren't supporting some extremely conservative OS that still ships an ancient pip.

Also, if we use `sys.executable` instead of hard-coding "python", it should work all the time, on every platform, venvs or not, even if somebody calls their interpreter something odd. Actually, let's do `sys.executable or 'python'`, because `executable` can be `None` or `''` in some corner cases.

Reply:
Yes, it was my approach in the previous iteration several months ago. But we did a fallback, as @bmw said, because the test-everything detected a regression if nearly EOL Linux distributions shipping really old pip versions.

However, we removed several of these distributions since then. So it may be actually working!

I will make the update here, and start an equivalent of test-everything.

Reply:
test-everything worked successfully